NorthstarDLL/shared/gamepresence.cpp

```diff
@@ -0,0 +1,81 @@
+#include "pch.h"
+
+#include "gamepresence.h"
+#include "plugins/pluginbackend.h"
+#include "plugins/plugins.h"
+#include "dedicated/dedicated.h"
+#include "server/serverpresence.h"
+#include "masterserver/masterserver.h"
+#include "squirrel/squirrel.h"
+
+GameStatePresence* g_pGameStatePresence;
+
+GameStatePresence::GameStatePresence()
+{
+	g_pServerPresence->AddPresenceReporter(&m_GameStateServerPresenceReporter);
+}
+
+void GameStateServerPresenceReporter::RunFrame(double flCurrentTime, const ServerPresence* pServerPresence)
+{
+	g_pGameStatePresence->id = pServerPresence->m_sServerId;
+	g_pGameStatePresence->name = pServerPresence->m_sServerName;
+	g_pGameStatePresence->description = pServerPresence->m_sServerDesc;
+	g_pGameStatePresence->password = pServerPresence->m_Password;
+
+	g_pGameStatePresence->map = pServerPresence->m_MapName;
+	g_pGameStatePresence->playlist = pServerPresence->m_PlaylistName;
+	g_pGameStatePresence->current_players = pServerPresence->m_iPlayerCount;
+	g_pGameStatePresence->max_players = pServerPresence->m_iMaxPlayers;
+
+	g_pGameStatePresence->is_local = !IsDedicatedServer();
+}
+
+void GameStatePresence::RunFrame()
+{
+	if (g_pSquirrel<ScriptContext::UI>->m_pSQVM != nullptr && g_pSquirrel<ScriptContext::UI>->m_pSQVM->sqvm != nullptr)
+		g_pSquirrel<ScriptContext::UI>->Call("GenerateUIPresence");
+
+	if (g_pSquirrel<ScriptContext::CLIENT>->m_pSQVM != nullptr && g_pSquirrel<ScriptContext::CLIENT>->m_pSQVM->sqvm != nullptr)
+	{
+		auto test = g_pSquirrel<ScriptContext::CLIENT>->Call("GenerateGameState");
+	}
+	g_pPluginCommunicationhandler->GeneratePresenceObjects();
+}
+
+ADD_SQFUNC("void", NSPushGameStateData, "struct gamestate", "", ScriptContext::CLIENT)
```

Comment:
not a fan of this, we need an actual struct definition rather than just taking any struct, if we're going to read values from memory that depend on it being the right struct type